### ISSUE TITLE
Running DeepQMC on multinode clusters

### ISSUE THREAD

**Issue by bkincaid256** [deepqmc/deepqmc#236]
Hello all,

I got the chance to speak with Zeno in person during the APS March meeting a couple weeks ago and had a good discussion regarding some of the less advertised capabilities of DeepQMC. The main one of interest to me at the moment would be running DeepQMC on multiple GPU nodes simultaneously, which Zeno mentioned was possible. I was wondering if it would be possible to see an example script, and get some explanation on what setup was needed to get it running parallelized across nodes. I'm specifically hoping to run on Perlmutter, but if there are any examples running on clusters with multiple distinct nodes coupled together, that would be a great starting point for me to fiddle with.

Sincerely,
Ben Kincaid

**Comment by zenoone** [comment]
Hi Ben,
thanks for the issue. Yes, running DeepQMC on multiple nodes of a compute cluster is possible, with one small caveat that I have just learned about. Having looked at the Perlmutter documentation, it seems that jobs can be scheduled with SLURM, which is great. 

DeepQMC relies on the JAX parallelisation capabilities. When DeepQMC is loaded in a multi-node setup, `jax.distributed.initialize()` is called automatically. This links the allocated devices with JAX's internal network, enabling JAX to access all the GPUs. ([deepqmc.\_\_init__](https://github.com/deepqmc/deepqmc/blob/bc9b15af45b0b9ba8c9c5d9e93410627d92bfe0e/src/deepqmc/__init__.py#L30) -> [deepqmc.parallel](https://github.com/deepqmc/deepqmc/blob/bc9b15af45b0b9ba8c9c5d9e93410627d92bfe0e/src/deepqmc/parallel.py#L39)) . However, there is a caveat: DeepQMC version 1.2.0 uses pmapping (rather than the more modern versions of sharding). In older versions of JAX (I believe <0.8.0), pmap can be run in a multi-node setup, whereas newer versions handle it differently, disabling pmap across multiple nodes. Make sure to downgrade JAX if you are running a newer version. 
```
 pip install --upgrade "jax[cuda12]==0.7.2" -f https://storage.googleapis.com/jax-releases/jax_cuda_releases.html
```
Here is an example how to run a simple training job on two nodes. 

For the job submission (`job.sh`):
```
#!/bin/bash
#SBATCH -J workdir
#SBATCH -o path/to/your/results/%j.out
#SBATCH --time=24:00:00
#SBATCH --partition=gpu
#SBATCH --nodes=2
#SBATCH --ntasks-per-node=1
#SBATCH --cpus-per-task=16
#SBATCH --gres=gpu:4
#SBATCH --exclude=

source /path/to/your/venv/bin/activate

srun bash /path/to/deepqmc_command
```
Ofc, the details of this config might depend on the specifics of the cluster and the job your running.

 `deepqmc_command` is just a plain text file containing the test command, i.e.:
```
deepqmc \
task=train \
hydra.run.dir=path/to/your/workdir \
ansatz=psiformer \
task.electron_batch_size=2000
```
Run the command with `sbatch job.sh`.  In your `workdir` DeepQMC should now generate one output directory per node (to avoid inter-node communication for logging). If you inspect the `deepqmc_0.log`, you should find lines similar to those:

```
[2026-04-07 12:41:24.722] INFO:deepqmc.app: Entering application
[2026-04-07 12:41:33.961] INFO:deepqmc.app: Process 0 running on gpu086
[2026-04-07 12:41:33.961] INFO:deepqmc.app: Running on 8 NVIDIA GEFORCE GTX 1080 TIs with 2 processes
```
This indicates that your multi-node setup is working. 

Let me know if you have any more questions or encounter any problems. I'm happy to help!

**Comment by bkincaid256** [comment]
Thanks for the example Zeno. I was able to get this running with the command line input with one modification, for whatever reason using jax 0.7.2 caused an error related to a removed some xla backend management. The error stated that a feature was removed in 0.7 so I rolled it back to jax 0.6 and that made things work as expected. 

I checked and this seems to work with the python script approach as well. I do have one more quick question though, is there a way to get the python script method to spit out a log file during the run like with the command line input? Using tensorboard logging isn't really convenient in my case.

**Comment by zenoone** [comment]
Great! Finding a workable combination of (up-to-date) packages can be a bit tricky at times, since the JAX ecosystem is evolving rapidly. 

Regarding logging to a file/terminal from a script, surely that is doable. For the event logging DeepQMC uses the [python logging module](https://docs.python.org/3/library/logging.html). If you run your script you most likely simply don't capture the log, which the command line interface (cli) handles automatically. One way to solve this is to add the following at the very beginning of your script:

 ```
import logging

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s - %(levelname)s - %(message)s",
    handlers=[
        logging.StreamHandler(),
        logging.FileHandler("path/to/your/logdir/deepqmc.log")
    ],
    force=True 
)
```

The `logging.StreamHandler()` redirects the output to the terminal from which you run the script and `logging.FileHandler("path/to/your/logdir/deepqmc.log")` writes the logs to file. 

There are different [logging levels](https://docs.python.org/3/library/logging.html#logging-levels) that you can configure by setting:

```
logging.basicConfig(
      level=logging.LEVEL,
      ....)
logging.getLogger('jax').setLevel(logging.LEVEL)
logging.getLogger('absl').setLevel(logging.LEVEL)
```

The default logging for the cli is:

```
logging.basicConfig(
      level=logging.DEBUG,
      ....)
logging.getLogger('jax').setLevel(logging.ERROR)
logging.getLogger('absl').setLevel(logging.ERROR)
```

As for the training progress, if you set a working directory in the train function
```
from deepqmc.train import train
train(hamil, ansatz, optimizer, sampler, workdir='path/to/workdir')
``` 
there will be both a tensorboard log file and a hdf5 log file created in the working directory. You can use 
```
import h5py
with h5py.File('path/to/workdir/training/result.h5', swmr=True) as f: ...
```
to read the file live while the training is progressing.

**Comment by bkincaid256** [comment]
Thanks for all the help with this. I tried everything out and it seems to be running no problem on Perlmutter.

**Comment by bkincaid256** [comment]
So I have been running some calculations for various systems, e.g. dimers and atoms, but noticed that the efficiency of training takes a nose dive compared to running on a single node. Is that expected behavior? For most conventional QMC methods, doubling the nodes while keeping the number of walkers and the system otherwise the same halves the wall time. In my experience, doing this with DeepQMC doesn't halve the run time as I would have expected, but reduces the wall time by roughly 20% or so compared to the single node equivalent.  I'll try to get some plots put together and put them in follow up comments.

That feels like like pretty bleak return on node hour investment, so is that expected or is there something going wrong with my installation specifically?

**Comment by zenoone** [comment]
Great that you got it to work! On our cluster we observe nearly linear strong scaling with the number of GPUs. 

<img width="340" height="340" alt="Image" src="https://github.com/user-attachments/assets/29ea8416-e1f9-4ba6-8829-9cc460effb96" />

Have you monitored the GPU throughput. In order for the scaling to be very efficient the GPUs need a certain level of saturation. Maybe try increasing the batch size and see if this changes the behavior (the batch is split across GPUs). 

In principle there should be just a small amount of communication, but if the connection between GPUs is very slow that can be a potential bottleneck. Are you running multiple GPUs on a single node or across nodes?

**Comment by bkincaid256** [comment]
With Perlmutter, its running on multiple nodes that each have 4 GPUs. I have stuck to mostly 2-4 node runs. I was using the a batch size of 2048, as I had seen used for a variety of system sizes. I could easily scale up the batch, but at that point the wall time would increase as well.

**Comment by zenoone** [comment]
I see, this is checking all the boxes. 

Firstly, I would try to understand why the parallelisation is underperforming. You could restrict yourself to using the GPUs within a single node and compare the results with using the same number of GPUs spread across nodes. This could provide some insight into an inter-node communication bottleneck. A more severe issue is probably the unsaturated GPU throughput. Since GPUs are good at performing large-scale linear algebra, they work best when matrix multiplication is the bottleneck. However, splitting your batch into smaller chunks (e.g. 2048/8 -> 256) moves you into a regime where each GPU is performing relatively small computations, meaning other parts of the code/computation may become the bottleneck (i.e. moving data around instead of crunching numbers). Sometimes, even the CPU can become a limiting factor, so make sure you allocate enough CPU resources when scaling up to many GPUs. 

If your primary objective is to reduce the walltime of the computation and you are in a regime where partitioning your batch further gives diminishing returns, the best strategy depends on your specific simulation task. If for instance you are interested in training multiple similar systems, it is clearly the most efficient to switch to the transferable setting [[1](https://arxiv.org/abs/2503.19847),[2](https://arxiv.org/abs/2603.25381)] and solve multiple geometries together. This also naturally scales to more GPUs. Conversely, if you are interested in a single molecule, you can increase the batch size alongside the learning rate https://github.com/deepqmc/deepqmc/blob/f9e1ff5ed0d4abe124c10d1a216d6ef27ca701fc/src/deepqmc/conf/task/opt/kfac.yaml#L6-L9 i.e. set it via `task.opt.kfac.learning_rate_schedule.init_value = 0.2`. 

Another very important optimization that will also increase the memory footprint (and therefore makes more GPUs useful or necessary) is using the [forward Laplacian](https://www.nature.com/articles/s42256-024-00794-x). In DeepQMC, we use [folx](https://github.com/microsoft/folx), which is an amazing implementation of the forward Laplacian for jax. Download it via `%pip install folx` and activate it by setting `hamil=qc_forward_laplacian` to obtain a speedup of around twofold without any approximations being made.

**Comment by bkincaid256** [comment]
I appreciate all the tips. I plan to test out the forward Laplacian module and see how it changes the rate of training. From what you have said, the drop in efficiency shouldn't be inherent, but is likely due to settings of the specific runs. I had been seeing this even with shared parameter runs, e.g. training one network on multiple geometries as shown in https://github.com/deepqmc/deepqmc/issues/231 , though again the partitioning of the walkers per card likely was the main culprit in diminishing returns. I'll see about trying a larger scale calculation and verifying how the efficiency measures up. Do you have a rule of thumb of how many walkers you think should be be on each card minimum to see efficient parallelization? 1024, 2048?

**Comment by zenoone** [comment]
For intermediate systems, I would try to keep the number of walkers per GPU to 2048, reducing it to 1024 or less only when OOM errors start to occur with larger systems. In the transferable setting, various training modes are possible. Some practitioners split the batch across tens or even hundreds of geometries, while others keep the training sequential, that is change the molecules only after a few iterations. Both approaches seem to work, and we typically choose a middle ground where we train on multiple geometries for a couple of steps and than swap the molecule batch with a new one. Considering your setup, I would try running 8 geometries with 2,048 walkers per geometry, or 16 geometries with 1,024. That said, there may be some ambiguity as to which is the optimal choice. The only constraint is that, in DeepQMC, your electron batch/number of walkers must be divisible by the number of GPUs.
